### PR TITLE
Expand 'opt_query' VOL callback to return a bitfield instead of a flag

### PR DESCRIPTION
**Pull request by qkoziol**
Allows querying more information from a VOL connector.

### CODE REVIEW

**Review by soumagne — APPROVED** [review]

**Review by derobins — APPROVED** [review]

**Review by byrnHDF — COMMENTED** [review]
H5Oget_comment is used in Java wrappers, but H5VLquery_optionalH5VLquery_optional is not implemented. Oversight when H5VLquery_optional was added but not to the wrappers?

**Comment by qkoziol** [comment]
> H5Oget_comment is used in Java wrappers, but H5VLquery_optionalH5VLquery_optional is not implemented. Oversight when H5VLquery_optional was added but not to the wrappers?

Yes, that seems to have been missed with the original H5VLquery_optional checking.  I'll see if I can update all the language wrappers (FORTRAN, C++, and Java) to check with H5VLquery_optional before calling H5Oget_comment (same as the tools do).

File: src/H5VLnative_introspect.c [hunk 1]
```diff
@@ -35,16 +35,192 @@
...
+                case H5VL_NATIVE_DATASET_CHUNK_READ:
+                    *flags |= H5VL_OPT_QUERY_READ_DATA;
+                    break;
+
+                case H5VL_NATIVE_DATASET_CHUNK_WRITE:
+                    *flags |= H5VL_OPT_QUERY_WRITE_DATA;
+                    break;
+
+                default:
+                    HGOTO_ERROR(H5E_ARGS, H5E_BADVALUE, FAIL, "unknown optional attribute operation")
```
Comment (bmribler) [inline src/H5VLnative_introspect.c]: Should this be dataset operation?

File: src/H5VLnative_introspect.c [hunk 2]
```diff
@@ -35,16 +35,192 @@
...
+                case H5VL_NATIVE_FILE_SET_LIBVER_BOUNDS:
+                case H5VL_NATIVE_FILE_GET_MIN_DSET_OHDR_FLAG:
+                case H5VL_NATIVE_FILE_SET_MIN_DSET_OHDR_FLAG:
+                case H5VL_NATIVE_FILE_GET_MPI_ATOMICITY:
+                case H5VL_NATIVE_FILE_SET_MPI_ATOMICITY:
+                case H5VL_NATIVE_FILE_POST_OPEN:
+                    break;
+
+                default:
+                    HGOTO_ERROR(H5E_ARGS, H5E_BADVALUE, FAIL, "unknown optional attribute operation")
```
Comment (bmribler) [inline src/H5VLnative_introspect.c]: ...and this

File: src/H5VLnative_introspect.c [hunk 3]
```diff
@@ -35,16 +35,192 @@
...
+                    *flags |= H5VL_OPT_QUERY_NO_ASYNC;
+                    break;
+
+                case H5VL_NATIVE_GROUP_GET_OBJINFO:
+                    *flags |= H5VL_OPT_QUERY_QUERY_METADATA;
+                    break;
+#endif /* H5_NO_DEPRECATED_SYMBOLS */
+
+                default:
+                    HGOTO_ERROR(H5E_ARGS, H5E_BADVALUE, FAIL, "unknown optional attribute operation")
```
Comment (bmribler) [inline src/H5VLnative_introspect.c]: ...and this

**Review by bmribler — APPROVED** [review]
Please see typo comments.

File: src/H5VLpublic.h [hunk 4]
```diff
@@ -49,6 +49,25 @@
...
+/* Flags to return from H5VLquery_optional API and 'opt_query' callbacks */
+/* Note: Operations which access multiple objects' data or metadata in a
+ *      container should be registered as file-level optional operations.
+ *      (e.g. "H5Dwrite_multi" takes a list of datasets to write data to, so
+ *      a VOL connector that implemented it should register it as an optional
+ *      file operation, and pass-through VOL connectors that are stacked above
+ *      the connector that registered it should assume that dataset elements
+ *      for _any_ dataset in the file could be written to)
+ */
+#define H5VL_OPT_QUERY_SUPPORTED       0x0001 /* VOL connector supports this operation */
```
Comment (byrnHDF) [inline src/H5VLpublic.h:61]: When wrappers are created, these may need to be added - at least to Java constants.

**Review by byrnHDF — APPROVED** [review]
Expect an issue for wrappers to be created.

**Comment by qkoziol** [comment]
> Please see typo comments.

Got 'em, thanks!

**Comment by qkoziol** [comment]
> Expect an issue for wrappers to be created.

I gave the wrapper situation a little more thought and think that the language wrappers are working correctly.   The language wrappers should just invoke the optional operation and allow it to succeed or fail, same as if an application invoked it directly.  The tools are different because they are trying to display information and they are the application, in this case.   So, I don't think any further changes are needed for the language wrappers and don't plan to update them.

**Review by lrknox — APPROVED** [review]

**Pull request merged**